src/compiler/Stage/InferTypes/AssignIds.elm

```diff
@@ -40,39 +40,52 @@ Output:
 -}
 
 import AST.Canonical as Canonical
+import AST.Common.Located as Located
 import AST.Common.Type as Type
 import AST.Typed as Typed
 import Common
 import Dict.Any
 import Stage.InferTypes.IdSource as IdSource exposing (IdSource)
 
 
-assignIds : Canonical.Expr -> ( Typed.Expr, IdSource )
-assignIds expr =
-    assignIdsWith IdSource.empty expr
+assignIds : Canonical.LocatedExpr -> ( Typed.LocatedExpr, IdSource )
+assignIds located =
+    assignIdsWith IdSource.empty located
 
 
-assignIdsWith : IdSource -> Canonical.Expr -> ( Typed.Expr, IdSource )
-assignIdsWith idSource expr =
+assignIdsWith : IdSource -> Canonical.LocatedExpr -> ( Typed.LocatedExpr, IdSource )
+assignIdsWith idSource located =
     let
-        wrap : IdSource -> Typed.Expr_ -> ( Typed.Expr, IdSource )
-        wrap idSource_ expr_ =
-            IdSource.one idSource_ (\id -> ( expr_, Type.Var id ))
+        ( located_, idSource_ ) =
+            assignIdsWithHelp idSource (Located.unwrap located)
     in
-    case expr of
+    {- Keep location, for error context -}
+    ( Located.map (\_ -> located_) located
```

Comment:
The `located` name can be renamed to `locatedCanonicalExpr`
The `located_` name probably needs to change to `typedExpr` as it's not really located
And then here we can write `Located.replaceWith typedExpr locatedCanonicalExpr` instead of the map